src/qutip_qip/qir.py

```diff
@@ -0,0 +1,231 @@
+                )
+
+        if op.name == "X":
+            builder.x(module.qubits[op.targets[0]])
+        elif op.name == "Y":
+            builder.y(module.qubits[op.targets[0]])
+        elif op.name == "Z":
+            builder.z(module.qubits[op.targets[0]])
+        elif op.name == "S":
+            builder.s(module.qubits[op.targets[0]])
+        elif op.name == "T":
+            builder.t(module.qubits[op.targets[0]])
+        elif op.name == "SNOT":
+            builder.h(module.qubits[op.targets[0]])
+        elif op.name in ("CNOT", "CX"):
+            builder.cx(
+                module.qubits[op.controls[0]], module.qubits[op.targets[0]]
+            )
+        elif op.name == "CZ":
+            builder.cz(
+                module.qubits[op.controls[0]], module.qubits[op.targets[0]]
+            )
+        elif op.name == "RX":
+            builder.rx(op.control_value, module.qubits[op.targets[0]])
+        elif op.name == "RY":
+            builder.ry(op.control_value, module.qubits[op.targets[0]])
+        elif op.name == "RZ":
+            builder.rz(op.control_value, module.qubits[op.targets[0]])
+        elif op.name in ("CRZ", "TOFFOLI"):
+            raise NotImplementedError(
+                "Decomposition of CRZ and Toffoli gates into base "
+                + "profile instructions is not yet implemented."
+            )
+        else:
+            raise ValueError(
+                f"Gate {op.name} not supported by the basic QIR builder, "
+                + "and may require a custom declaration."
+            )
+
+    fmt = QirFormat.ensure(format)
+
+    module = pqg.SimpleModule(module_name, circuit.N, circuit.num_cbits or 0)
+    builder = pqg.BasicQisBuilder(module.builder)
+
+    for op in circuit.gates:
+        # If we have a QuTiP gate, then we need to convert it into one of
+        # the reserved operation names in the QIR base profile's quantum
+        # instruction set (QIS).
+        if isinstance(op, Gate):
+            # TODO: Validate indices.
+            append_operation(module, builder, op)
+
+        elif isinstance(op, Measurement):
+            builder.m(
+                module.qubits[op.targets[0]],
+                module.results[op.classical_store],
+            )
+
+        else:
+            raise NotImplementedError(
+                f"Instruction {op} is not implemented in the QIR base "
+                + "profile and may require a custom declaration."
+            )
+
+    if fmt == QirFormat.TEXT:
+        return module.ir()
+    elif fmt == QirFormat.BITCODE:
+        return module.bitcode()
+    elif fmt == QirFormat.MODULE:
+        bitcode = module.bitcode()
+        f = NamedTemporaryFile(suffix=".bc", delete=False)
+        try:
+            f.write(bitcode)
+        finally:
+            f.close()
+        module = pqp.QirModule(f.name)
+        try:
+            os.unlink(f.name)
+        except:
+            pass
```

Comment:
A small question, why is this `unlink` needed here?

Reply:
Ah, this ensures that the temporary file written in the previous step is deleted when we're done with it. Normally, that's handled when the `TemporaryFile` object gets garbage collected, but since we need to write to it after it's been closed, that obligates us to manually unlink it when we're done.